File: examples/backends/typescript/src/decimal.ts
/**
 * Exact decimal arithmetic on decimal *strings*.
 *
 * The Meteroid TypeScript SDK types every `format: decimal` value as a `string`, because
 * a JS `number` cannot hold one losslessly — and it ships no arithmetic for them. The
 * quota check has to subtract and compare those values, so this does it on `bigint`
 * scaled integers. Nothing in this backend ever passes a decimal through `Number()`:
 * that is how a 0.1-minute clip eventually bills wrong.
 */

import { ApiError } from "./error.js";

const PATTERN = /^-?\d+(\.\d+)?$/;

interface Scaled {
  units: bigint;
  scale: number;
}

function parse(value: string): Scaled {
  // The `typeof` is not paranoia: the SDK's deserializers do not validate, so a decimal
  // Meteroid sent as a JSON number arrives here as a `number` typed `string`. Every
  // decimal this backend parses came from Meteroid, so a malformed one is an upstream
  // problem rather than a bug here or a bad request.
  if (typeof value !== "string" || !PATTERN.test(value)) {
    throw new ApiError(
      "UPSTREAM_ERROR",
      `Meteroid returned ${JSON.stringify(value)} where an exact decimal was expected.`,
    );
  }
  const [whole = "", fraction = ""] = value.split(".");
  const negative = whole.startsWith("-");
  const units = BigInt(`${whole.replace("-", "")}${fraction}`);
  return { units: negative ? -units : units, scale: fraction.length };
}

function align(a: Scaled, b: Scaled): [bigint, bigint, number] {
  const scale = Math.max(a.scale, b.scale);
  const lift = (value: Scaled) => value.units * 10n ** BigInt(scale - value.scale);
  return [lift(a), lift(b), scale];
}

/** Trailing zeros trimmed, never scientific notation — what the contract's `Decimal` wants. */
function render({ units, scale }: Scaled): string {
  const negative = units < 0n;
  const digits = (negative ? -units : units).toString().padStart(scale + 1, "0");
  const whole = digits.slice(0, digits.length - scale);
  const fraction = digits.slice(digits.length - scale).replace(/0+$/, "");
  return `${negative ? "-" : ""}${whole}${fraction ? `.${fraction}` : ""}`;
}

/** Render a Meteroid decimal the way the contract requires: `"3.50"` becomes `"3.5"`. */
export function normalize(value: string): string {
  return render(parse(value));
}

export function normalizeOpt(value: string | null | undefined): string | null {
  return value == null ? null : normalize(value);
}

export function add(a: string, b: string): string {
  const [x, y, scale] = align(parse(a), parse(b));
  return render({ units: x + y, scale });
}

export function subtract(a: string, b: string): string {
  const [x, y, scale] = align(parse(a), parse(b));
  return render({ units: x - y, scale });
}

/** Numeric, not textual: `"1"` and `"1.00"` are the same balance. */
export function greaterThan(a: string, b: string): boolean {
  const [x, y] = align(parse(a), parse(b));
  return x > y;
}

/**
 * `durationSeconds / 60`, rounded **up** to two decimals — the demo always bills at
 * least what it used. Integer ceiling division, so `100 / 60` is exactly `1.67`.
 */
export function billableMinutes(durationSeconds: number): string {
  const hundredths = (BigInt(durationSeconds) * 100n + 59n) / 60n;
  return render({ units: hundredths, scale: 2 });
}

File: examples/backends/go/decimal.go
package main

import (
	"math/big"
	"regexp"
	"strings"
)

// Exact decimal arithmetic on decimal *strings*.
//
// The Meteroid Go SDK types every `format: decimal` value as a string, because a
// float64 cannot hold one losslessly — and, depending on the standard library only, it
// ships no arithmetic for them. The quota check has to subtract and compare those
// values, so this does it on big.Int scaled integers. Nothing in this backend ever
// passes a decimal through strconv.ParseFloat: that is how a 0.1-minute clip eventually
// bills wrong.

var decimalPattern = regexp.MustCompile(`^-?\d+(\.\d+)?$`)

// decimal is units × 10^-scale.
type decimal struct {
	units *big.Int
	scale int
}

// parseDecimal only ever sees values that came from Meteroid, so a malformed one is an
// upstream problem rather than a bug here or a bad request.
func parseDecimal(value string) (decimal, error) {
	if !decimalPattern.MatchString(value) {
		return decimal{}, newAPIError(codeUpstreamError,
			"Meteroid returned %q where an exact decimal was expected.", value)
	}
	whole, fraction, _ := strings.Cut(value, ".")
	// The pattern has already vouched for every digit, so SetString cannot fail.
	units, _ := new(big.Int).SetString(whole+fraction, 10)
	return decimal{units: units, scale: len(fraction)}, nil
}

// align lifts both operands to the larger of the two scales.
func align(a, b decimal) (x, y *big.Int, scale int) {
	scale = max(a.scale, b.scale)
	lift := func(d decimal) *big.Int {
		factor := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(scale-d.scale)), nil)
		return factor.Mul(factor, d.units)
	}
	return lift(a), lift(b), scale
}

// String renders the way the contract's `Decimal` wants: trailing zeros trimmed, never
// scientific notation, and no negative zero.
func (d decimal) String() string {
	digits := new(big.Int).Abs(d.units).String()
	if len(digits) <= d.scale {
		digits = strings.Repeat("0", d.scale-len(digits)+1) + digits
	}
	whole := digits[:len(digits)-d.scale]
	fraction := strings.TrimRight(digits[len(digits)-d.scale:], "0")

	var b strings.Builder
	if d.units.Sign() < 0 {
		b.WriteByte('-')
	}
	b.WriteString(whole)
	if fraction != "" {
		b.WriteByte('.')
		b.WriteString(fraction)
	}
	return b.String()
}

// normalizeDecimal renders a Meteroid decimal the way the contract requires:
// "3.50" becomes "3.5".
func normalizeDecimal(value string) (string, error) {
	d, err := parseDecimal(value)
	if err != nil {
		return "", err
	}
	return d.String(), nil
}

func normalizeDecimalOpt(value *string) (*string, error) {
	if value == nil {
		return nil, nil
	}
	normalized, err := normalizeDecimal(*value)
	if err != nil {
		return nil, err
	}
	return &normalized, nil
}

func addDecimal(a, b string) (string, error) {
	return combine(a, b, (*big.Int).Add)
}

func subtractDecimal(a, b string) (string, error) {
	return combine(a, b, (*big.Int).Sub)
}

func combine(a, b string, op func(z, x, y *big.Int) *big.Int) (string, error) {
	left, err := parseDecimal(a)
	if err != nil {
		return "", err
	}
	right, err := parseDecimal(b)
	if err != nil {
		return "", err
	}
	x, y, scale := align(left, right)
	return decimal{units: op(new(big.Int), x, y), scale: scale}.String(), nil
}

// decimalGreaterThan is numeric, not textual: "1" and "1.00" are the same balance.
func decimalGreaterThan(a, b string) (bool, error) {
	left, err := parseDecimal(a)
	if err != nil {
		return false, err
	}
	right, err := parseDecimal(b)
	if err != nil {
		return false, err
	}
	x, y, _ := align(left, right)
	return x.Cmp(y) > 0, nil
}

// billableMinutes is durationSeconds / 60, rounded **up** to two decimals — the demo
// always bills at least what it used. Integer ceiling division, so 100 / 60 is exactly
// "1.67".
func billableMinutes(durationSeconds int32) string {
	hundredths := (int64(durationSeconds)*100 + 59) / 60
	return decimal{units: big.NewInt(hundredths), scale: 2}.String()
}
